contracts/dapp/SelfHostedApplicationFactory.sol

```diff
@@ -0,0 +1,97 @@
+// (c) Cartesi and individual authors (see AUTHORS)
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity ^0.8.8;
+
+import {IConsensus} from "../consensus/IConsensus.sol";
+import {Authority} from "../consensus/authority/Authority.sol";
+import {IAuthorityFactory} from "../consensus/authority/IAuthorityFactory.sol";
+import {Application} from "./Application.sol";
+import {IApplicationFactory} from "./IApplicationFactory.sol";
+import {ISelfHostedApplicationFactory} from "./ISelfHostedApplicationFactory.sol";
+import {IInputBox} from "../inputs/IInputBox.sol";
+import {IPortal} from "../portals/IPortal.sol";
+
+/// @title Self-hosted Application Factory
+/// @notice Allows anyone to reliably deploy a new Authority contract,
+/// along with an Application contract already linked to it.
+contract SelfHostedApplicationFactory is ISelfHostedApplicationFactory {
+    IAuthorityFactory immutable _authorityFactory;
+    IApplicationFactory immutable _applicationFactory;
+
+    /// @param authorityFactory The authority factory
+    /// @param applicationFactory The application factory
+    constructor(
+        IAuthorityFactory authorityFactory,
+        IApplicationFactory applicationFactory
+    ) {
+        _authorityFactory = authorityFactory;
+        _applicationFactory = applicationFactory;
+    }
+
+    function getAuthorityFactory()
+        external
+        view
+        override
+        returns (IAuthorityFactory)
+    {
+        return _authorityFactory;
+    }
+
+    function getApplicationFactory()
+        external
+        view
+        override
+        returns (IApplicationFactory)
+    {
+        return _applicationFactory;
+    }
+
+    function deployContracts(
+        address authorityOwner,
+        IInputBox inputBox,
+        IPortal[] memory portals,
+        address appOwner,
+        bytes32 templateHash,
+        bytes32 salt
+    ) external returns (Application, Authority) {
+        Authority authority = _authorityFactory.newAuthority(
+            authorityOwner,
+            salt
+        );
+
+        Application application = _applicationFactory.newApplication(
+            authority,
+            inputBox,
+            portals,
+            appOwner,
+            templateHash,
+            salt
+        );
+
+        return (application, authority);
```

Comment:
You could make the return values named, just like you did in `calculateAddresses`, and just assign the newly-created contracts to them. With this, you won't need to declare them, won't need the `return` statement, and the first statement might even fit in one line?

Reply:
Hey Gui. I modified the `calculateAddresses` so there's no suffix `_` for return variables. I think suffix `_` for return variables make sense if there's prefix `_` for function parameters.  But we are using prefix `_` as internal state variables.

Reply:
Oh, I didn't notice the `_` suffix. Yeah, we're no longer doing that. But I think we can still name return values, like `application` and `authority`.

Reply:
True. Done. 👍